src/preamble_minimal.js

```diff
@@ -111,34 +72,33 @@ function updateGlobalBufferAndViews(b) {
   HEAPF32 = new Float32Array(b);
   HEAPF64 = new Float64Array(b);
 }
-updateGlobalBufferAndViews(buffer);
-#else
-// In non-ALLOW_MEMORY_GROWTH scenario, we only need to initialize
-// the heap once, so optimize code size to do it statically here.
-var HEAP8 = new Int8Array(buffer);
-var HEAP16 = new Int16Array(buffer);
-var HEAP32 = new Int32Array(buffer);
-var HEAPU8 = new Uint8Array(buffer);
-var HEAPU16 = new Uint16Array(buffer);
-var HEAPU32 = new Uint32Array(buffer);
-var HEAPF32 = new Float32Array(buffer);
-var HEAPF64 = new Float64Array(buffer);
-#endif
 
-#if USE_PTHREADS && ((MEM_INIT_METHOD == 1 && !MEM_INIT_IN_WASM && !SINGLE_FILE) || USES_DYNAMIC_ALLOC)
+#if EXTERNAL_MEMORY
 if (!ENVIRONMENT_IS_PTHREAD) {
+#if ALLOW_MEMORY_GROWTH && MAXIMUM_MEMORY != -1
+  var wasmMaximumMemory = {{{ MAXIMUM_MEMORY >>> 16 }}};
+#else
+  var wasmMaximumMemory = {{{ INITIAL_MEMORY >>> 16}}};
 #endif
-
-#if MEM_INIT_METHOD == 1 && !MEM_INIT_IN_WASM && !SINGLE_FILE
-#if ASSERTIONS
-if (!Module['mem']) throw 'Must load memory initializer as an ArrayBuffer in to variable Module.mem before adding compiled output .js script to the DOM';
-#endif
-HEAPU8.set(new Uint8Array(Module['mem']), {{{ GLOBAL_BASE }}});
+  wasmMemory = new WebAssembly.Memory({
+    'initial': {{{ INITIAL_MEMORY >>> 16 }}}
+#if USE_PTHREADS || !ALLOW_MEMORY_GROWTH || MAXIMUM_MEMORY != -1
+    , 'maximum': wasmMaximumMemory
 #endif
+    , 'shared': true
+    });
+}
 
-#if USE_PTHREADS && ((MEM_INIT_METHOD == 1 && !MEM_INIT_IN_WASM && !SINGLE_FILE) || USES_DYNAMIC_ALLOC)
+#if MODULARIZE
+if (ENVIRONMENT_IS_WORKER) {
+  updateGlobalBufferAndViews({{{EXPORT_NAME}}}.buffer);
+} else {
+  updateGlobalBufferAndViews(wasmMemory.buffer);
```

Comment:
This `if (ENVIRONMENT_IS_WORKER)` looks odd and like a code size regression? Is this intended to be the case for pthread local thread init? Then would the following work?

```js
#if USE_PTHREADS
updateGlobalBufferAndViews({{{EXPORT_NAME}}}.buffer || wasmMemory.buffer);
#else
updateGlobalBufferAndViews(wasmMemory.buffer);
#endif
```

Reply:
Oh now I see it comes from the moved block below. Lgtm. (looks like there is room to optimize this later)